File: benchmarks/scripts/summarize_results.py
#!/usr/bin/env python3
"""Aggregate hyperfine benchmark JSON results into summary tables.

Usage:
    python3 summarize_results.py <results-dir> [--json <output-path>]

Output:
    Markdown summary table to stdout.
    JSON summary to file if --json <path> is specified.

Reads:
    *-quick.json, *-full.json, *-stress.json   hyperfine result files
    *-subsystems.json                           Zig subsystem timing files (if present)
    metric-accuracy.json                        compare_metrics.py output (if present)

Hyperfine JSON schema:
    { results: [
        { command, mean, stddev, memory_usage_byte: [int, ...], ... },  // CG
        { command, mean, stddev, memory_usage_byte: [int, ...], ... },  // FTA
    ]}
"""

import json
import os
import sys
import glob
from pathlib import Path


def mean_memory(memory_list) -> float:
    """Compute mean memory usage in MB from a list of bytes values."""
    if not memory_list:
        return 0.0
    return sum(memory_list) / len(memory_list) / (1024 * 1024)


def parse_hyperfine_file(filepath: str) -> dict | None:
    """Parse a single hyperfine JSON result file.

    Returns dict with project, suite, cg_mean_ms, cg_stddev_ms, fta_mean_ms,
    fta_stddev_ms, cg_mem_mb, fta_mem_mb, speedup, mem_ratio, or None on error.
    """
    filename = os.path.basename(filepath)
    # Extract project and suite from filename: <project>-<suite>.json
    parts = filename.replace(".json", "").rsplit("-", 1)
    if len(parts) != 2:
        return None
    project, suite = parts[0], parts[1]

    try:
        with open(filepath) as f:
            data = json.load(f)
    except (json.JSONDecodeError, FileNotFoundError) as e:
        print(f"Warning: Could not load {filepath}: {e}", file=sys.stderr)
        return None

    results = data.get("results", [])
    if len(results) < 2:
        # Some files may have only one result if FTA failed; skip
        if len(results) == 1:
            print(f"Warning: {filename} has only 1 result (expected 2), skipping", file=sys.stderr)
        return None

    cg = results[0]
    fta = results[1]

    cg_mean_ms = cg.get("mean", 0.0) * 1000
    cg_stddev_ms = cg.get("stddev", 0.0) * 1000
    fta_mean_ms = fta.get("mean", 0.0) * 1000
    fta_stddev_ms = fta.get("stddev", 0.0) * 1000

    cg_mem_list = cg.get("memory_usage_byte", [])
    fta_mem_list = fta.get("memory_usage_byte", [])
    cg_mem_mb = mean_memory(cg_mem_list)
    fta_mem_mb = mean_memory(fta_mem_list)

    # speedup > 1.0 means FTA is faster (CG takes longer relative to FTA)
    speedup = cg_mean_ms / fta_mean_ms if fta_mean_ms > 0 else 0.0
    mem_ratio = fta_mem_mb / cg_mem_mb if cg_mem_mb > 0 else 0.0

    return {
        "project": project,
        "suite": suite,
        "cg_mean_ms": round(cg_mean_ms, 1),
        "cg_stddev_ms": round(cg_stddev_ms, 1),
        "fta_mean_ms": round(fta_mean_ms, 1),
        "fta_stddev_ms": round(fta_stddev_ms, 1),
        "cg_mem_mb": round(cg_mem_mb, 1),
        "fta_mem_mb": round(fta_mem_mb, 1),
        "speedup": round(speedup, 2),
        "mem_ratio": round(mem_ratio, 2),
    }


def parse_subsystems_file(filepath: str) -> dict | None:
    """Parse a Zig subsystem benchmark JSON file.

    Expected schema: { project, suite, subsystems: [{ name, mean_ms, stddev_ms }] }
    Returns None on error.
    """
    try:
        with open(filepath) as f:
            data = json.load(f)
    except (json.JSONDecodeError, FileNotFoundError) as e:
        print(f"Warning: Could not load {filepath}: {e}", file=sys.stderr)
        return None
    return data


def parse_metric_accuracy(filepath: str) -> list | None:
    """Parse metric-accuracy.json produced by compare-metrics.sh.

    Expected schema: [{ project, files_compared, cyclomatic, halstead_volume, ... }]
    Returns None on error.
    """
    try:
        with open(filepath) as f:
            data = json.load(f)
    except (json.JSONDecodeError, FileNotFoundError):
        return None
    return data if isinstance(data, list) else None


def format_speed_row(r: dict) -> str:
    """Format a markdown table row for the speed comparison table.

    speedup = cg_mean / fta_mean: >1.0 means FTA is faster than CG.
    """
    speedup_str = f"{r['speedup']:.1f}x"
    if r["speedup"] > 1.0:
        # FTA is faster
        speedup_label = f"{speedup_str} faster"
    elif r["speedup"] < 0.99:
        # CG is faster
        speedup_label = f"{1/r['speedup']:.1f}x CG faster"
    else:
        speedup_label = "equal"

    cg_ms_str = f"{r['cg_mean_ms']:.0f} ± {r['cg_stddev_ms']:.0f}"
    fta_ms_str = f"{r['fta_mean_ms']:.0f} ± {r['fta_stddev_ms']:.0f}"

    return (
        f"| {r['project']:<20} | {cg_ms_str:<12} | {fta_ms_str:<12} | "
        f"{speedup_label:<12} | {r['cg_mem_mb']:<10.1f} | {r['fta_mem_mb']:<11.1f} | "
        f"{r['mem_ratio']:.1f}x      |"
    )


def print_speed_table(results: list[dict], suite: str | None = None) -> None:
    """Print a markdown speed and memory comparison table."""
    filtered = [r for r in results if suite is None or r["suite"] == suite]
    if not filtered:
        return

    suite_label = suite or "all"
    print(f"\n### Speed and Memory Comparison ({suite_label} suite)\n")
    print(f"| {'Project':<20} | {'CG (ms)':<12} | {'FTA (ms)':<12} | {'Speedup':<12} | {'CG Mem (MB)':<10} | {'FTA Mem (MB)':<11} | Mem Ratio |")
    print(f"| {'-'*20} | {'-'*12} | {'-'*12} | {'-'*12} | {'-'*10} | {'-'*11} | {'-'*9} |")
    for r in sorted(filtered, key=lambda x: x["cg_mean_ms"]):
        print(format_speed_row(r))

    if len(filtered) > 1:
        avg_speedup = sum(r["speedup"] for r in filtered) / len(filtered)
        avg_mem_ratio = sum(r["mem_ratio"] for r in filtered if r["mem_ratio"] > 0) / max(
            sum(1 for r in filtered if r["mem_ratio"] > 0), 1
        )
        print(f"\n**Mean speedup:** {avg_speedup:.1f}x &nbsp; | &nbsp; **Mean memory ratio:** {avg_mem_ratio:.1f}x")
        print(f"*(Speedup = CG time / FTA time; >1.0 means FTA is faster than CG)*")


def print_metric_accuracy_table(accuracy_data: list) -> None:
    """Print a markdown metric accuracy summary table."""
    if not accuracy_data:
        return

    print("\n### Metric Accuracy: CG vs FTA Agreement\n")
    print("| Project | Files | Cyclo Agree | Cyclo Corr | Halstead Agree | Halstead Corr |")
    print("| ------- | ----- | ----------- | ---------- | -------------- | ------------- |")
    for item in accuracy_data:
        project = item.get("project", "?")
        n = item.get("files_compared", 0)
        cyclo = item.get("cyclomatic", {})
        hal = item.get("halstead_volume", {})
        print(
            f"| {project:<20} | {n:<5} | "
            f"{cyclo.get('within_tolerance_pct', 0):.0f}% (±25%)  | "
            f"{cyclo.get('ranking_correlation', 0):.3f}      | "
            f"{hal.get('within_tolerance_pct', 0):.0f}% (±30%)      | "
            f"{hal.get('ranking_correlation', 0):.3f}         |"
        )

    # Overall averages
    if accuracy_data:
        avg_cyclo = sum(d.get("cyclomatic", {}).get("within_tolerance_pct", 0) for d in accuracy_data) / len(accuracy_data)
        avg_hal = sum(d.get("halstead_volume", {}).get("within_tolerance_pct", 0) for d in accuracy_data) / len(accuracy_data)
        print(f"\n**Mean cyclomatic agreement:** {avg_cyclo:.0f}% &nbsp; | &nbsp; **Mean Halstead agreement:** {avg_hal:.0f}%")


def main():
    if len(sys.argv) < 2:
        print("Usage: summarize_results.py <results-dir> [--json <output-path>]", file=sys.stderr)
        sys.exit(1)

    results_dir = sys.argv[1]
    json_output_path = None

    # Parse --json flag
    args = sys.argv[2:]
    i = 0
    while i < len(args):
        if args[i] == "--json" and i + 1 < len(args):
            json_output_path = args[i + 1]
            i += 2
        else:
            i += 1

    if not os.path.isdir(results_dir):
        print(f"Error: results directory not found: {results_dir}", file=sys.stderr)
        sys.exit(1)

    # Discover hyperfine result files
    suite_pattern = os.path.join(results_dir, "*-quick.json")
    suite_files = (
        glob.glob(os.path.join(results_dir, "*-quick.json"))
        + glob.glob(os.path.join(results_dir, "*-full.json"))
        + glob.glob(os.path.join(results_dir, "*-stress.json"))
    )

    benchmark_results = []
    for filepath in sorted(suite_files):
        r = parse_hyperfine_file(filepath)
        if r is not None:
            benchmark_results.append(r)

    if not benchmark_results:
        print("Warning: No hyperfine benchmark result files found.", file=sys.stderr)

    # Discover subsystem files
    subsystem_files = glob.glob(os.path.join(results_dir, "*-subsystems.json"))
    subsystem_data = []
    for filepath in sorted(subsystem_files):
        d = parse_subsystems_file(filepath)
        if d is not None:
            subsystem_data.append(d)

    # Load metric accuracy if present
    accuracy_path = os.path.join(results_dir, "metric-accuracy.json")
    accuracy_data = parse_metric_accuracy(accuracy_path)

    # Determine which suites are present
    suites = sorted(set(r["suite"] for r in benchmark_results))

    # Print markdown output
    print("## ComplexityGuard vs FTA: Benchmark Summary\n")
    print(f"Results from: `{results_dir}`\n")

    for suite in suites:
        print_speed_table(benchmark_results, suite=suite)

    if len(suites) > 1:
        print_speed_table(benchmark_results, suite=None)

    # Subsystem breakdown
    if subsystem_data:
        print("\n### CG Subsystem Breakdown\n")
        # Collect all subsystem names across projects
        all_subsystems = set()
        for d in subsystem_data:
            for s in d.get("subsystems", []):
                all_subsystems.add(s["name"])
        subsystem_names = sorted(all_subsystems)
        header = "| Project | " + " | ".join(subsystem_names) + " |"
        separator = "| ------- | " + " | ".join("---" for _ in subsystem_names) + " |"
        print(header)
        print(separator)
        for d in subsystem_data:
            subs = {s["name"]: s.get("mean_ms", 0) for s in d.get("subsystems", [])}
            row = f"| {d.get('project', '?'):<20} | " + " | ".join(
                f"{subs.get(n, 0):.1f} ms" for n in subsystem_names
            ) + " |"
            print(row)

    # Metric accuracy
    if accuracy_data:
        print_metric_accuracy_table(accuracy_data)
    else:
        print("\n*Metric accuracy data not found. Run `compare-metrics.sh` to generate it.*")

    # Overall summary stats
    if benchmark_results:
        avg_speedup = sum(r["speedup"] for r in benchmark_results) / len(benchmark_results)
        avg_mem = sum(r["mem_ratio"] for r in benchmark_results if r["mem_ratio"] > 0)
        n_mem = sum(1 for r in benchmark_results if r["mem_ratio"] > 0)
        avg_mem_ratio = avg_mem / max(n_mem, 1)
        fastest_cg = min(benchmark_results, key=lambda r: r["cg_mean_ms"])
        slowest_cg = max(benchmark_results, key=lambda r: r["cg_mean_ms"])
        print(f"\n### Overall Summary\n")
        print(f"- Projects benchmarked: {len(benchmark_results)}")
        print(f"- Mean CG/FTA speed ratio: {avg_speedup:.2f}x (>1.0 means FTA is faster than CG)")
        print(f"- Mean FTA/CG memory ratio: {avg_mem_ratio:.2f}x (FTA uses more memory = >1.0)")
        print(f"- Fastest project: {fastest_cg['project']} ({fastest_cg['cg_mean_ms']:.0f} ms CG)")
        print(f"- Slowest project: {slowest_cg['project']} ({slowest_cg['cg_mean_ms']:.0f} ms CG)")

    # JSON output
    if json_output_path:
        summary = {
            "results_dir": results_dir,
            "benchmark_results": benchmark_results,
            "subsystem_data": subsystem_data,
            "accuracy_data": accuracy_data,
        }
        os.makedirs(os.path.dirname(os.path.abspath(json_output_path)), exist_ok=True)
        with open(json_output_path, "w") as f:
            json.dump(summary, f, indent=2)
        print(f"\n*JSON summary written to: {json_output_path}*", file=sys.stderr)


if __name__ == "__main__":
    main()

File: benchmarks/scripts/summarize-results.mjs
#!/usr/bin/env node
/**
 * Aggregate hyperfine benchmark JSON results into summary tables.
 *
 * Usage:
 *   node summarize-results.mjs <results-dir> [--json <output-path>]
 *
 * Output:
 *   Markdown summary table to stdout.
 *   JSON summary to file if --json <path> is specified.
 *
 * Reads:
 *   *-quick.json, *-full.json, *-stress.json   hyperfine result files
 *   *-subsystems.json                           subsystem timing files (if present)
 *
 * Hyperfine JSON schema:
 *   { results: [
 *     { command, mean, stddev, memory_usage_byte: [int, ...], ... },  // CG
 *   ]}
 */

import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';

/**
 * Compute mean memory usage in MB from a list of bytes values.
 * @param {number[]} memoryList
 * @returns {number}
 */
function meanMemory(memoryList) {
  if (!memoryList || memoryList.length === 0) return 0.0;
  const total = memoryList.reduce((sum, v) => sum + v, 0);
  return total / memoryList.length / (1024 * 1024);
}

/**
 * Round a number to a given number of decimal places.
 * @param {number} value
 * @param {number} decimals
 * @returns {number}
 */
function round(value, decimals) {
  const factor = Math.pow(10, decimals);
  return Math.round(value * factor) / factor;
}

/**
 * Parse a single hyperfine JSON result file.
 *
 * Returns object with project, suite, cg_mean_ms, cg_stddev_ms, cg_mem_mb,
 * or null on error.
 * @param {string} filepath
 * @returns {object|null}
 */
function parseHyperfineFile(filepath) {
  const filename = path.basename(filepath);
  // Extract project and suite from filename: <project>-<suite>.json
  const nameNoExt = filename.replace(/\.json$/, '');
  const lastDash = nameNoExt.lastIndexOf('-');
  if (lastDash === -1) return null;
  const project = nameNoExt.slice(0, lastDash);
  const suite = nameNoExt.slice(lastDash + 1);

  let data;
  try {
    const content = fs.readFileSync(filepath, 'utf8');
    data = JSON.parse(content);
  } catch (e) {
    process.stderr.write(`Warning: Could not load ${filepath}: ${e.message}\n`);
    return null;
  }

  const results = data.results || [];
  if (results.length < 1) {
    return null;
  }

  const cg = results[0];

  const cgMeanMs = (cg.mean || 0.0) * 1000;
  const cgStddevMs = (cg.stddev || 0.0) * 1000;

  const cgMemList = cg.memory_usage_byte || [];
  const cgMemMb = meanMemory(cgMemList);

  return {
    project,
    suite,
    cg_mean_ms: round(cgMeanMs, 1),
    cg_stddev_ms: round(cgStddevMs, 1),
    cg_mem_mb: round(cgMemMb, 1),
  };
}

/**
 * Parse a CG analysis JSON output file.
 *
 * Extracts file count, function count, metric averages, and health score.
 * @param {string} filepath
 * @returns {object|null}
 */
function parseAnalysisFile(filepath) {
  try {
    const content = fs.readFileSync(filepath, 'utf8');
    const data = JSON.parse(content);
    const summary = data.summary || {};
    const files = data.files || [];

    // Compute average metrics across all functions
    let totalCyclomatic = 0;
    let totalCognitive = 0;
    let totalHalstead = 0;
    let funcCount = 0;

    for (const file of files) {
      for (const fn of file.functions || []) {
        totalCyclomatic += fn.cyclomatic || 0;
        totalCognitive += fn.cognitive || 0;
        totalHalstead += fn.halstead_volume || 0;
        funcCount++;
      }
    }

    const project = path.basename(filepath).replace(/-analysis\.json$/, '');

    return {
      project,
      files_analyzed: summary.files_analyzed || 0,
      total_functions: summary.total_functions || funcCount,
      warnings: summary.warnings || 0,
      errors: summary.errors || 0,
      health_score: round(summary.health_score || 0, 1),
      avg_cyclomatic: funcCount > 0 ? round(totalCyclomatic / funcCount, 1) : 0,
      avg_cognitive: funcCount > 0 ? round(totalCognitive / funcCount, 1) : 0,
      avg_halstead: funcCount > 0 ? round(totalHalstead / funcCount, 0) : 0,
    };
  } catch (e) {
    return null;
  }
}

/**
 * Parse a subsystem benchmark JSON file.
 *
 * Handles two schemas:
 *   - Array schema: { project, subsystems: [{ name, mean_ms, stddev_ms }] }
 *   - Object schema: { project, subsystems: { name: { mean_ms, ... }, ... } }
 * Normalizes to array schema before returning. Returns null on error.
 * @param {string} filepath
 * @returns {object|null}
 */
function parseSubsystemsFile(filepath) {
  try {
    const content = fs.readFileSync(filepath, 'utf8');
    const data = JSON.parse(content);
    // Normalize object-keyed subsystems to array format
    if (data.subsystems && !Array.isArray(data.subsystems) && typeof data.subsystems === 'object') {
      data.subsystems = Object.entries(data.subsystems).map(([name, stats]) => ({
        name,
        mean_ms: stats.mean_ms || 0,
        stddev_ms: stats.stddev_ms || 0,
      }));
    }
    return data;
  } catch (e) {
    process.stderr.write(`Warning: Could not load ${filepath}: ${e.message}\n`);
    return null;
  }
}

/**
 * Pad a string to a given width (left-aligned).
 * @param {string|number} value
 * @param {number} width
 * @returns {string}
 */
function padEnd(value, width) {
  return String(value).padEnd(width);
}

/**
 * Format a markdown table row for the performance table.
 * @param {object} r
 * @param {object|null} analysis
 * @returns {string}
 */
function formatSpeedRow(r, analysis) {
  const cgMsStr = `${Math.round(r.cg_mean_ms)} \u00b1 ${Math.round(r.cg_stddev_ms)}`;

  if (analysis) {
    return (
      `| ${padEnd(r.project, 20)} | ${padEnd(analysis.files_analyzed, 5)} | ${padEnd(analysis.total_functions, 5)} | ${padEnd(cgMsStr, 14)} | ${padEnd(r.cg_mem_mb.toFixed(1), 8)} | ${padEnd(analysis.avg_cyclomatic, 4)} | ${padEnd(analysis.avg_cognitive, 4)} | ${padEnd(analysis.avg_halstead, 7)} | ${padEnd(analysis.warnings, 4)} | ${padEnd(analysis.errors, 4)} | ${padEnd(analysis.health_score, 5)} |`
    );
  }

  return (
    `| ${padEnd(r.project, 20)} | ${padEnd('-', 5)} | ${padEnd('-', 5)} | ${padEnd(cgMsStr, 14)} | ${padEnd(r.cg_mem_mb.toFixed(1), 8)} | ${padEnd('-', 4)} | ${padEnd('-', 4)} | ${padEnd('-', 7)} | ${padEnd('-', 4)} | ${padEnd('-', 4)} | ${padEnd('-', 5)} |`
  );
}

/**
 * Print a markdown performance table.
 * @param {object[]} results
 * @param {string|null} suite
 * @param {Map<string, object>} analysisMap
 */
function printSpeedTable(results, suite = null, analysisMap = new Map()) {
  const filtered = suite === null ? results : results.filter(r => r.suite === suite);
  if (filtered.length === 0) return;

  const suiteLabel = suite || 'all';
  console.log(`\n### Performance (${suiteLabel} suite)\n`);
  console.log(`| ${padEnd('Project', 20)} | ${padEnd('Files', 5)} | ${padEnd('Funcs', 5)} | ${padEnd('CG (ms)', 14)} | ${padEnd('Mem (MB)', 8)} | ${padEnd('Cyc', 4)} | ${padEnd('Cog', 4)} | ${padEnd('Halsted', 7)} | ${padEnd('Warn', 4)} | ${padEnd('Err', 4)} | ${padEnd('Score', 5)} |`);
  console.log(`| ${'-'.repeat(20)} | ${'-'.repeat(5)} | ${'-'.repeat(5)} | ${'-'.repeat(14)} | ${'-'.repeat(8)} | ${'-'.repeat(4)} | ${'-'.repeat(4)} | ${'-'.repeat(7)} | ${'-'.repeat(4)} | ${'-'.repeat(4)} | ${'-'.repeat(5)} |`);

  const sorted = [...filtered].sort((a, b) => a.cg_mean_ms - b.cg_mean_ms);
  for (const r of sorted) {
    const analysis = analysisMap.get(r.project) || null;
    console.log(formatSpeedRow(r, analysis));
  }

  if (filtered.length > 1) {
    const avgMs = filtered.reduce((sum, r) => sum + r.cg_mean_ms, 0) / filtered.length;
    const withMem = filtered.filter(r => r.cg_mem_mb > 0);
    const avgMem = withMem.length > 0
      ? withMem.reduce((sum, r) => sum + r.cg_mem_mb, 0) / withMem.length
      : 0;

    // Compute analysis averages
    const withAnalysis = filtered.filter(r => analysisMap.has(r.project));
    let analysisStats = '';
    if (withAnalysis.length > 0) {
      const avgFiles = Math.round(withAnalysis.reduce((sum, r) => sum + (analysisMap.get(r.project)?.files_analyzed || 0), 0) / withAnalysis.length);
      const avgFuncs = Math.round(withAnalysis.reduce((sum, r) => sum + (analysisMap.get(r.project)?.total_functions || 0), 0) / withAnalysis.length);
      const avgScore = round(withAnalysis.reduce((sum, r) => sum + (analysisMap.get(r.project)?.health_score || 0), 0) / withAnalysis.length, 1);
      analysisStats = ` \u00a0 | \u00a0 **Mean files:** ${avgFiles} \u00a0 | \u00a0 **Mean functions:** ${avgFuncs} \u00a0 | \u00a0 **Mean health score:** ${avgScore}`;
    }
    console.log(`\n**Mean analysis time:** ${Math.round(avgMs)} ms \u00a0 | \u00a0 **Mean memory:** ${avgMem.toFixed(1)} MB${analysisStats}`);
  }
}

function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    process.stderr.write('Usage: summarize-results.mjs <results-dir> [--json <output-path>]\n');
    process.exit(1);
  }

  const resultsDir = args[0];
  let jsonOutputPath = null;

  // Parse --json flag
  for (let i = 1; i < args.length; i++) {
    if (args[i] === '--json' && i + 1 < args.length) {
      jsonOutputPath = args[i + 1];
      i++;
    }
  }

  if (!fs.existsSync(resultsDir) || !fs.statSync(resultsDir).isDirectory()) {
    process.stderr.write(`Error: results directory not found: ${resultsDir}\n`);
    process.exit(1);
  }

  // Discover hyperfine result files
  const allFiles = fs.readdirSync(resultsDir).map(f => path.join(resultsDir, f));
  const suiteFiles = allFiles.filter(f =>
    f.endsWith('-quick.json') || f.endsWith('-normal.json') || f.endsWith('-full.json') || f.endsWith('-stress.json')
  ).sort();

  const benchmarkResults = [];
  for (const filepath of suiteFiles) {
    const r = parseHyperfineFile(filepath);
    if (r !== null) benchmarkResults.push(r);
  }

  if (benchmarkResults.length === 0) {
    process.stderr.write('Warning: No hyperfine benchmark result files found.\n');
  }

  // Discover analysis files
  const analysisFiles = allFiles.filter(f => f.endsWith('-analysis.json')).sort();
  const analysisMap = new Map();
  for (const filepath of analysisFiles) {
    const a = parseAnalysisFile(filepath);
    if (a !== null) analysisMap.set(a.project, a);
  }

  // Discover subsystem files
  const subsystemFiles = allFiles.filter(f => f.endsWith('-subsystems.json')).sort();
  const subsystemData = [];
  for (const filepath of subsystemFiles) {
    const d = parseSubsystemsFile(filepath);
    if (d !== null) subsystemData.push(d);
  }

  // Determine which suites are present
  const suiteSet = new Set(benchmarkResults.map(r => r.suite));
  const suites = [...suiteSet].sort();

  // Load system info if present
  const systemInfoPath = path.join(resultsDir, 'system-info.json');
  let systemInfo = null;
  try {
    systemInfo = JSON.parse(fs.readFileSync(systemInfoPath, 'utf8'));
  } catch (_e) { /* system-info.json not present in older baselines */ }

  // Print markdown output
  console.log('## ComplexityGuard Benchmark Summary\n');
  console.log(`Results from: \`${resultsDir}\`\n`);

  if (systemInfo) {
    console.log('### System\n');
    console.log('| Component | Value |');
    console.log('| --------- | ----- |');
    console.log(`| CPU | ${systemInfo.cpu.model} (${systemInfo.cpu.cores} cores / ${systemInfo.cpu.threads} threads) |`);
    console.log(`| Memory | ${systemInfo.memory.total_gb} GB |`);
    console.log(`| OS | ${systemInfo.os} (kernel ${systemInfo.kernel}) |`);
    console.log(`| Architecture | ${systemInfo.arch} |`);
    console.log('');
  }

  for (const suite of suites) {
    printSpeedTable(benchmarkResults, suite, analysisMap);
  }

  if (suites.length > 1) {
    printSpeedTable(benchmarkResults, null, analysisMap);
  }

  // Subsystem breakdown
  if (subsystemData.length > 0) {
    console.log('\n### CG Subsystem Breakdown\n');
    // Collect all subsystem names across projects
    const allSubsystems = new Set();
    for (const d of subsystemData) {
      for (const s of d.subsystems || []) {
        allSubsystems.add(s.name);
      }
    }
    const subsystemNames = [...allSubsystems].sort();
    const header = '| Project | ' + subsystemNames.join(' | ') + ' |';
    const separator = '| ------- | ' + subsystemNames.map(() => '---').join(' | ') + ' |';
    console.log(header);
    console.log(separator);
    for (const d of subsystemData) {
      const subs = {};
      for (const s of d.subsystems || []) {
        subs[s.name] = s.mean_ms || 0;
      }
      const row = `| ${padEnd(d.project || '?', 20)} | ` +
        subsystemNames.map(n => `${(subs[n] || 0).toFixed(1)} ms`).join(' | ') + ' |';
      console.log(row);
    }
  }

  // Overall summary stats
  if (benchmarkResults.length > 0) {
    const avgMs = benchmarkResults.reduce((sum, r) => sum + r.cg_mean_ms, 0) / benchmarkResults.length;
    const withMem = benchmarkResults.filter(r => r.cg_mem_mb > 0);
    const avgMem = withMem.length > 0
      ? withMem.reduce((sum, r) => sum + r.cg_mem_mb, 0) / withMem.length
      : 0;
    const fastestCg = benchmarkResults.reduce((min, r) => r.cg_mean_ms < min.cg_mean_ms ? r : min);
    const slowestCg = benchmarkResults.reduce((max, r) => r.cg_mean_ms > max.cg_mean_ms ? r : max);

    console.log('\n### Overall Summary\n');
    console.log(`- Projects benchmarked: ${benchmarkResults.length}`);
    console.log(`- Mean analysis time: ${Math.round(avgMs)} ms`);
    console.log(`- Mean memory usage: ${avgMem.toFixed(1)} MB`);
    console.log(`- Fastest project: ${fastestCg.project} (${Math.round(fastestCg.cg_mean_ms)} ms)`);
    console.log(`- Slowest project: ${slowestCg.project} (${Math.round(slowestCg.cg_mean_ms)} ms)`);
  }

  // JSON output
  if (jsonOutputPath) {
    const summary = {
      results_dir: resultsDir,
      system_info: systemInfo,
      benchmark_results: benchmarkResults,
      analysis_data: Object.fromEntries(analysisMap),
      subsystem_data: subsystemData,
    };
    const dir = path.dirname(path.resolve(jsonOutputPath));
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(jsonOutputPath, JSON.stringify(summary, null, 2));
    process.stderr.write(`\n*JSON summary written to: ${jsonOutputPath}*\n`);
  }
}

main();
